trie/branches.py

```diff
@@ -0,0 +1,190 @@
+    the prefix of an existing key in the trie.
+    """
+    validate_is_bytes(key_prefix)
+
+    return _if_branch_exist(db, root_hash, encode_to_bin(key_prefix))
+
+
+def _if_branch_exist(db, node_hash, key_prefix):
+    # Empty trie
+    if node_hash == BLANK_HASH:
+        return False
+    nodetype, left_child, right_child = parse_node(db[node_hash])
+    if nodetype == LEAF_TYPE:
+        if key_prefix:
+            return False
+        return True
+    elif nodetype == KV_TYPE:
+        if not key_prefix:
+            return True
+        if len(key_prefix) < len(left_child):
+            if key_prefix == left_child[:len(key_prefix)]:
+                return True
+            return False
+        else:
+            if key_prefix[:len(left_child)] == left_child:
+                return _if_branch_exist(db, right_child, key_prefix[len(left_child):])
+            return False
+    elif nodetype == BRANCH_TYPE:
+        if not key_prefix:
+            return True
+        if key_prefix[:1] == BYTE_0:
+            return _if_branch_exist(db, left_child, key_prefix[1:])
+        else:
+            return _if_branch_exist(db, right_child, key_prefix[1:])
+    else:
+        raise Exception("Invariant: unreachable code path")
+
+
+def get_branch(db, root_hash, key):
+    """
+    Get a long-format Merkle branch
+    """
+    validate_is_bytes(key)
+
+    return list(_get_branch(db, root_hash, encode_to_bin(key)))
+
+
+def _get_branch(db, node_hash, keypath):
+    if node_hash == BLANK_HASH:
+        raise StopIteration
+    node = db[node_hash]
+    nodetype, left_child, right_child = parse_node(node)
+    if nodetype == LEAF_TYPE:
+        if not keypath:
+            yield node
+        else:
+            raise InvalidKeyError("Key too long")
+    elif nodetype == KV_TYPE:
+        if not keypath:
+            raise InvalidKeyError("Key too short")
+        if keypath[:len(left_child)] == left_child:
+            yield node
+            yield from _get_branch(
+                db,
+                right_child,
+                keypath[len(left_child):]
+            )
+        else:
+            yield node
+    elif nodetype == BRANCH_TYPE:
+        if not keypath:
+            raise InvalidKeyError("Key too short")
+        if keypath[:1] == BYTE_0:
+            yield node
+            yield from _get_branch(db, left_child, keypath[1:])
+        else:
+            yield node
+            yield from _get_branch(db, right_child, keypath[1:])
+    else:
+        raise Exception("Invariant: unreachable code path")
+
+
+def if_branch_valid(branch, root_hash, key, value):
+    # value being None means it's not in the trie
+    if value is not None:
+        validate_is_bytes(key)
+    for node in branch:
+        validate_is_bin_node(node)
+
+    db = {keccak(node): node for node in branch}
+    assert BinaryTrie(db=db, root_hash=root_hash).get(key) == value
+    return True
+
+
+def get_trie_nodes(db, node_hash):
```

Comment:
It's worth noting that this function will return a generator which is probably not what you actually want.  I would suggest writing an outer function `get_trie_nodes` and renaming this function to `_get_trie_nodes` like the following.


```python
def get_trie_nodes(...):
    return tuple(_get_trie_nodes(...))
```

This will allow all of the various iterators to be lazily evaluated at the very end and cast to a single tuple.